galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+    @param screen_scale  Physical pixel scale of phase screen in meters.  A fraction of the Fried
+                         parameter is usually sufficiently small, but users should test the effects
+                         of this parameter to ensure robust results. [Default: half of r0_500]
+    @param altitude      Altitude of phase screen in km.  This is with respect to the telescope, not
+                         sea-level.  [Default: 0.0]
+    @param time_step     Interval to use when advancing the screen in time in seconds.
+                         [Default: 0.03]
+    @param r0_500        Fried parameter setting the amplitude of turbulence; contributes to "size"
+                         of the resulting atmospheric PSF.  Specified at wavelength 500 nm, in units
+                         of meters.  [Default: 0.2]
+    @param L0_inv        Inverse outer scale in inverse meters, beyond which turbulence power
+                         spectrum smoothly approaches zero.  [Default: 1./25]
+    @param vx            x-component wind velocity in meters/second.  [Default: 0.]
+    @param vy            y-component wind velocity in meters/second.  [Default: 0.]
+    @param alpha_mag     Fraction of phase that is "remembered" between time_steps.  The fraction
+                         1-alpha_mag is then the amount of turbulence freshly generated in each
+                         step.  [Default: 0.997]
+    @param rng           Random number generator as a galsim.BaseDeviate().  If None, then use the
+                         clock time or system entropy to seed a new generator.  [Default: None]
+
+    Relevant SPIE paper:
+    "Remembrance of phases past: An autoregressive method for generating realistic atmospheres in
+    simulations"
+    Srikar Srinath, Univ. of California, Santa Cruz;
+    Lisa A. Poyneer, Lawrence Livermore National Lab.;
+    Alexander R. Rudy, UCSC; S. Mark Ammons, LLNL
+    Published in Proceedings Volume 9148: Adaptive Optics Systems IV
+    September 2014
+    """
+    def __init__(self, screen_size, screen_scale, altitude=0.0, time_step=0.03,
+                 r0_500=0.2, L0_inv=1./25.0, vx=0.0, vy=0.0, alpha_mag=0.997, rng=None):
+
+        super(ARAtmosphericScreen, self).__init__(screen_size, screen_scale, altitude,
+                                                  time_step, r0_500, L0_inv,
+                                                  vx, vy, rng)
+
+        self.alpha_mag = alpha_mag
+        fx, fy = self._freq()
+        self._init_screen(fx, fy)
+
+        # Work wind directly into "boiling" phase updating.
+        alpha_phase = -(fx*vx + fy*vy) * time_step
+        self.alpha = alpha_mag * np.exp(2j*np.pi*alpha_phase)
+
+        self.noise_frac = np.sqrt(1.0 - np.abs(self.alpha**2)[0, 0])
+
+        # Ignore boiling for *really* large alpha, but don't let amplitudes decay
+        if self.noise_frac < 1.e-12:
+            self.alpha /= np.abs(self.alpha)
```

Comment:
>What is the point of this? Is there a rounding error issue that alpha is sometimes essentially 1, but not quite, and things go wrong if it's not put back to be exactly |alpha| = 1?

I can't recall what I had in mind here, but this disappeared in the refactor of ARAtmosphericScreen + FrozenAtmosphericScreen -> AtmosphericScreen anyway.